### PR TITLE
csync: Use Qt for encoding conversion instead of iconv

### PR DESCRIPTION
**Pull request by ogoffart**
Replaces #5715

Should fix issues:

- #5661   On mac, iconv did not support all of unicode and some files with emoji in the filename could not be uploaded
    
- #5719 , #5676  On linux, we will now support non utf-8 locale

### CODE REVIEW

**Comment by ckamm** [comment]
This looks good, in particular the dependency removal and being able to use Qt in csync.

I worry about testing a bit - if this introduces issues we'll only notice by seeing bad filenames on users' machines. We should run the encoding tests on all platforms (I assume you covered linux?), and maybe check how thorough they are.

File: csync/src/std/c_utf8.cc [hunk 1]
```diff
@@ -64,26 +60,33 @@ char* c_utf8_from_locale(const mbchar_t *wstr)
     WideCharToMultiByte(CP_UTF8, 0, wstr, len, mdst, size_needed, NULL, NULL);
     dst = mdst;
   }
+  return dst;
 #else
-  dst = c_strdup(wstr);
+    QTextDecoder dec(QTextCodec::codecForLocale());
+    QString s = dec.toUnicode(wstr, qstrlen(wstr));
```
Comment (dragotin) [inline csync/src/std/c_utf8.cc]: `const QString s = ....` ?

Tested on Linux and works, but might be an issue on MacOSX

Comment (ogoffart) [inline csync/src/std/c_utf8.cc]: 's' can't be const for two reason:  First, on mac, we update it with `s = s.normalized(QString::NormalizationForm_C);`  
Second, we std::move it before calling toUtf8. This saves an allocation as there is an overload of toUtf8 which does the conversion in-place. (but does not work if s was const)

**Comment by ogoffart** [comment]
Regarding the  check_encoding.c, it is true that the #ifdef HAVE_ICONV will not be compiled and could be cleaned up.  I just kept it because i did not want to change the test. It should anyway pass with or without iconv as the behavior should be the same. (except for the codepoints that were not understood by macos's iconv)


The CI is failiong for unrelated reason (it ran out of disk)

Any objection for merging this?

**Comment by ogoffart** [comment]
@jturcotte quickly tested this patch on windows and did not notice regressions.

Since there are no objections, i'm merging now.

**Comment by guruz** [comment]
Some troubles with APFS see here: https://github.com/owncloud/client/issues/5650
(But they appear unrelated to this PR..)

**Pull request merged**